// apollo-angular-boost/esm5/ApolloBoost.js
import { __assign, __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { ApolloLink, Observable } from 'apollo-link';
import { Apollo } from 'apollo-angular';
import { HttpLink } from 'apollo-angular-link-http';
import { withClientState } from 'apollo-link-state';
import { onError } from 'apollo-link-error';
import { InMemoryCache } from 'apollo-cache-inmemory';
var ApolloBoost = /** @class */ (function () {
    function ApolloBoost(apollo, httpLink) {
        this.apollo = apollo;
        this.httpLink = httpLink;
    }
    ApolloBoost.prototype.create = function (config) {
        var cache = config && config.cacheRedirects
            ? new InMemoryCache({ cacheRedirects: config.cacheRedirects })
            : new InMemoryCache();
        var stateLink = config && config.clientState
            ? withClientState(__assign(__assign({}, config.clientState), { cache: cache }))
            : false;
        var errorLink = config && config.onError
            ? onError(config.onError)
            : onError(function (_a) {
                var graphQLErrors = _a.graphQLErrors, networkError = _a.networkError;
                if (graphQLErrors) {
                    graphQLErrors.map(function (_a) {
                        var message = _a.message, locations = _a.locations, path = _a.path;
                        // tslint:disable-next-line
                        return console.log("[GraphQL error]: Message: " + message + ", Location: " + locations + ", Path: " + path);
                    });
                }
                if (networkError) {
                    // tslint:disable-next-line
                    console.log("[Network error]: " + networkError);
                }
            });
        var requestHandler = config && config.request
            ? new ApolloLink(function (operation, forward) {
                return new Observable(function (observer) {
                    var handle;
                    Promise.resolve(operation)
                        .then(function (oper) { return config.request(oper); })
                        .then(function () {
                        handle = forward(operation).subscribe({
                            next: observer.next.bind(observer),
                            error: observer.error.bind(observer),
                            complete: observer.complete.bind(observer),
                        });
                    })
                        .catch(observer.error.bind(observer));
                    return function () {
                        if (handle) {
                            handle.unsubscribe();
                        }
                    };
                });
            })
            : false;
        var httpLink = this.httpLink.create(__assign(__assign({}, ((config && config.httpOptions) || {})), { uri: (config && config.uri) || '/graphql' }));
        var link = ApolloLink.from([
            errorLink,
            requestHandler,
            stateLink,
            httpLink,
        ].filter(function (x) { return !!x; }));
        this.apollo.create({
            link: link,
            cache: cache,
        });
    };
    ApolloBoost.ctorParameters = function () { return [
        { type: Apollo },
        { type: HttpLink }
    ]; };
    ApolloBoost = __decorate([
        Injectable()
    ], ApolloBoost);
    return ApolloBoost;
}());
export { ApolloBoost };
